### ISSUE TITLE
Be careful with accessibility

### ISSUE THREAD

**Issue by stephanie-walter** [tawian/text-spinners#6]
Hello,
I just opened the demo using voice over and some buttons are really annoying. It will read the emojis, but also the arrows and a few other symbols "loading heart green heart yellow heart red heart green heart yellow ..." is quite annoying. 
I've not found a solution yet, but maybe come accessibility experts might know how to solve this?

**Comment by maxbeier** [comment]
Hi,

wow, good point. Didn't think about that.

I added a note about accessibility and suggest the usage of the ARIA attributes `aria-hidden` or `aria-label`. It's not perfect and should be adapted to the actual spinner in use, but better than it was before. What do you think about that?

**Comment by cookiecrook** [comment]
Since these are progress indicators, add the following ARIA to each of your markup examples:

``` html
role="progressbar" aria-label="Loading…"
```

**Comment by maxbeier** [comment]
Ah, ok. Changed it to:

> For better accessibility with screen readers add `aria-hidden="true"` or `role="progressbar" aria-label="Loading…"` to the loading element.

Is that correct or should the `role` be on the element even when it's hidden?

**Comment by cookiecrook** [comment]
That's better, but people are unlikely to read your disclaimer. The are more likely to copy your markup examples without reading:

``` html
<span class="loading dots"></span>
```

You could solve this by using a role attr selector instead of your functional classname:

``` html
<span role="progressbar" class="dots"></span>
```

Then your CSS could use attr selectors as a backwards-compatible alias to your existing functional classname ".loading":

``` css
[role="progressbar"], .loading { /* default styles */ }
[role="progressbar"].dots, .loading.dots { /* braille dots */ }
[role="progressbar"].dots2, .loading.dots2 { /* braille dots 2 */ }
```

So your stylesheet would still work for all the existing implementations, but you've updated your suggested markup example to be more accessible.

@maxbeier wrote:

> I suggest the usage of `aria-hidden`

That's okay, but unnecessary of they end up hiding it with `display:none;` or `visibility:hidden;`

> should the `role` be on the element even when it's hidden?

It's harmless to have a role on a hidden element.

**Comment by maxbeier** [comment]
Hi,

I'm not experienced with screen readers or ARIA, but if I only add `role="progressbar"`, voice over on macOS doesn't change (I assume because it's just a hint on what this element would look like if you could see it?). Only `aria-hidden` and `aria-label` seem to make a concrete difference (reading nothing or the provided label).

So at the moment I don't see the point in changing the syntax if there is no measurable effect and

``` html
<!-- with text -->
<button>Loading <span class="loading dots" aria-hidden="true"></span></button>
```

or 

``` html
<!-- only the spinner -->
<button>Loading <span class="loading hearts" aria-label="Loading…"></span></button>
```

seem to provide the best experience (not reading the dots or the hearts).

Please correct me if I'm not getting the point (really).

Greets
Max

**Comment by cookiecrook** [comment]
`<button>` is a leaf node, so all contents are flattened. The markup makes a difference if it is used outside of a button.

``` html
<span role="progressbar" aria-valuetext="Loading…" aria-valuenow="25" aria-valuemin="0" aria-valuemax="100">...</span>
```

**Comment by maxbeier** [comment]
Is the current state ok for you? I would close the issue then as the initial problem is solved / mitigated from my point of view.
